lisp/x/profiler/funlabeler.go

```diff
@@ -0,0 +1,85 @@
+package profiler
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/luthersystems/elps/lisp"
+	"github.com/luthersystems/elps/lisp/lisplib/libhelp"
+)
+
+// FunLabeler provides an alternative name for a function label in the trace.
+type FunLabeler func(runtime *lisp.Runtime, fun *lisp.LVal) string
+
+// WithELPSDocLabeler labels spans using elps doc magic strings.
+func WithELPSDocLabeler() Option {
+	return WithFunLabeler(elpsDocFunLabeler)
+}
+
+// WithFunLabeler sets the labeler for tracing spans.
+func WithFunLabeler(funLabeler FunLabeler) Option {
+	return func(p *profiler) {
+		p.funLabeler = funLabeler
+	}
+}
+
+// defaultFunName constructs a pretty canonical name using the function name.
+func defaultFunName(runtime *lisp.Runtime, fun *lisp.LVal) string {
+	if fun.Type != lisp.LFun {
+		return ""
+	}
+	funData := fun.FunData()
+	if funData == nil {
+		return ""
+	}
+	return fmt.Sprintf("%s:%s", funData.Package, getFunNameFromFID(runtime, funData.FID))
+}
+
+// ELPSDocLabel is a magic string used to extract function labels.
+const ELPSDocLabel = `@trace{([^}]+)}`
+
+var elpsDocLabelRegExp = regexp.MustCompile(ELPSDocLabel)
+
+var sanitizeRegExp = regexp.MustCompile(`[\W_]+`)
+
+func sanitizeLabel(userLabel string) string {
+	userLabel = strings.TrimSpace(userLabel)
+	if userLabel == "" {
+		return ""
+	}
+
+	userLabel = sanitizeRegExp.ReplaceAllString(userLabel, "_")
+	// Ensure the label doesn't start with a digit or special character
+	if len(userLabel) > 0 && !unicode.IsLetter(rune(userLabel[0])) {
+		userLabel = "label_" + userLabel
+	}
+
+	// Eliminate duplicate underscores
+	parts := strings.Split(userLabel, "_")
+	var nonEmptyParts []string
+	for _, part := range parts {
+		if part != "" {
+			nonEmptyParts = append(nonEmptyParts, part)
+		}
+	}
+	return strings.Join(nonEmptyParts, "_")
```

Comment:
This prefixing behavior seems a bit non-intuitive.  Would it make sense to just drop leading junk we don't want to allow or reject the label entirely and fall back to the function name?